### PR TITLE
feat(gateway): Add admin session authentication manager

### LINKED ISSUE #4128: Use session tokens for admin API auth
Users would acquire a session token through a login API endpoint. This token would stay valid for a limited time and be sent with every API request like the password is currently.

This allows:
* More elegant login attempt rate limiting (#4122)
* 2FA login

### PR DESCRIPTION
**Pull request by f3r10**
closes #4128 
# Motivation
Currently, authentication to the getaway admin endpoints is made through a password sent in clear text, which is a security concern.
This PR attempts to fix this issue by adding an Authentication Manager that distributes challenge codes that must be signed before users can request a JWT that allows them to authenticate to the gateway admin endpoints. 
Because of backward compatibility, the password check is not deleted but used as a fallback only if a JWT is absent in the request.
# This PR adds three additional endpoints:
- `/auth/challenge` generates a temporal challenge code
- `/auth/sign` once the challenge code is generated, this endpoint signs the challenge code
- `/auth/session` Once the challenge code is signed, it is possible to request a JWT token.
# How to test it:
- run `just mprocs`
- check which gateway address has to be used  running for example `gateway-lnd info`
- check which is the federation ID that has to be used with some commands such as `balance`
image
- Be aware that the process of requesting a challenge code, signing it, and requesting a JWT has to be done in less than one minute because of the expiration time.
- the JWT has an expiration time of 30 minutes
- `gateway-cli --address=http://127.0.0.1:13446/v1 get-challenge-auth`
- `gateway-cli --address=http://127.0.0.1:13446/v1 sign-challenge-auth --challenge-auth="02f0219bdc2ea9b4524a78f0a390863e16967940e9048171994b73d81305d76d7d-http://127.0.0.1:13446/v1-1728495691"`
- `gateway-cli --address=http://127.0.0.1:13446/v1 session-auth --challenge="02f0219bdc2ea9b4524a78f0a390863e16967940e9048171994b73d81305d76d7d-http://127.0.0.1:13446/v1-1728495691" --response="3acaa8ac174da093ff43eaab77e7ff916c202c8d744c2ec41df42f347b47ab03f162bb8c2a092008e804c90f9251d4553b95d8e2305d00e14012e19a25da6cb5"`
- `gateway-cli --address=http://127.0.0.1:13446/v1 --rpcjwt="eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6ImlkLTY1MDQ5MDMxMDUzNzQxMzYzNTUiLCJleHAiOjE3Mjg0OTU2OTF9.mvBJcOLQqIZy5J8IDSr6SE9g9wSD0MiXbV9o6J9-30c" balance`

### CODE REVIEW

**Review by elsirion — COMMENTED** [review]
Just some comments from a quick skim. Feels like work in progress, will mark it as draft.

File: gateway/ln-gateway/src/auth_manager.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        let gateway_id = PublicKey::from_str(parts[0])?;
+        let gateway_api = parts[1].to_string();
+        let expiry = parts[2].parse::<u64>()?;
+        Ok(Self {
+            gateway_id,
+            gateway_api,
+            expiry,
+        })
+    }
+}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply gateway/ln-gateway/src/auth_manager.rs:153]: this is for the challenge not for the actual jwt, it uses that crate there.

**Comment by f3r10** [comment]
> Just some comments from a quick skim. Feels like work in progress, will mark it as draft.

Thanks, @elsirion for the initial review, I addressed the comments and the PR will be ready for additional reviews.

**Review by otech47 — COMMENTED** [review]
tested this and works as expected based on PR example steps

Screenshot 2024-11-08 at 11 41 44 AM

I also tested that jwt session keys still work after generating a 2nd one. I generated 2 session keys and could use either of them to hit the balance

**Comment by elsirion** [comment]
@elsirion review

**Comment by Kodylow** [comment]
> tested this and works as expected based on PR example steps
> 
> Screenshot 2024-11-08 at 11 41 44 AM
> I also tested that jwt session keys still work after generating a 2nd one. I generated 2 session keys and could use either of them to hit the balance

the session keys just expire based on time, do we want to keep it to just a single one active ever?

**Comment by m1sterc001guy** [comment]
> the session keys just expire based on time, do we want to keep it to just a single one active ever?

I think its fine to have two, maybe you want to access it from multiple devices. Can we rebase and run the CI?

**Comment by f3r10** [comment]
there are some errors on the `Build on linux` CI task that I am fixing now.

**Comment by f3r10** [comment]
closing this PR for a simple version #6535

**Comment by f3r10** [comment]
close this PR for #6535

**Pull request closed without merging**